Draft of 3.0 prerelease 2 blog post (#1197)
lit/lit.dev: packages/lit-dev-content/site/blog/2023-09-27-lit-3.0-prerelease-2.md

---
tags: blogPosts
layout: blog-post.html
title: "Lit 3.0 Prerelease 2 and more!"
summary: "Today we are publishing a second of prereleases for the upcoming Lit 3.0 launch."
date: 2023-09-27
author:
  - justin-fagnani
---

# Lit 3.0 Prerelease 2 and more!

Today we are publishing a second of prereleases for the upcoming Lit 3.0 launch. This includes the core packages of `lit`, `lit-html`, and `lit-element`, prereleases for the Task, Context, and React packages that are graduating from Lit Labs, and previews of two exciting new Lit Labs packages: `@lit-labs/compiler` and `@lit-labs/preact-signals`.

The full list of prerelease packages is:
- `lit@3.0.0-pre.1`
- `lit-html@3.0.0-pre.1`
- `lit-element@4.0.0-pre.1`
- `@lit/reactive-element@3.0.0-pre.1`
- `@lit/context@1.0.0-pre.1`
- `@lit/task@1.0.0-pre.1`
- `@lit/react@1.0.0-pre.1`

> - `@lit/task@1.0.0-pre.1`

Reviewer: task duplicated in list

- `@lit-labs/compiler@0.1.0-pre.1`
- `@lit-labs/preact-signals@0.1.0-pre.1`

## Lit 3.0 Prerelease 2

As we [announced with the first prerelease](./2023-05-15-lit-3.0-prerelease/), Lit 3.0 is a breaking change that most notably removes support for IE11. Lit 3.0 is intended to have no new features, since new features can be added in non-breaking minor versions.

### New hybrid decorators with standard decorators support

> One new thing we _have_ added to Lit 3.0 though is support for TC39 standard decorators. The new decorator spec has reached Stage 3 in TC39, meaning that browsers and compilers are now implementing them. This is a huge step for Lit because it allows us to begin the process of moving to decorator implementation that won't require a compiler to use.

Reviewer 2: to decorator -> to a decorator

> One new thing we _have_ added to Lit 3.0 though is support for TC39 standard decorators. The new decorator spec has reached Stage 3 in TC39, meaning that browsers and compilers are now implementing them. This is a huge step for Lit because it allows us to begin the process of moving to decorator implementation that won't require a compiler to use.

Reviewer: ```suggestion
One new thing we _have_ added to Lit 3.0 though is support for TC39 standard decorators. The new decorator spec has reached Stage 3 in TC39, meaning that browsers and compilers are now implementing them. This is a huge step for Lit because it allows us to begin the process of moving to a decorator implementation that won't require a compiler to use.
```

> Recently TypeScript 5.2 added support for standard decorator metadata, which was the last piece needed for us to use the new specification (Babel's next release will have support). So we've made our decorators work with TypeScript's `experimentalDecorators` flag set to either `true` or `false`.

Reviewer 2: Looks like Babel just released v7.23.0 with metadata: https://babeljs.io/blog/2023/09/25/7.23.0

> The reason why we've included standard decorator support in a breaking change is that we've taken great care to make the upgrade path from experimental decorators smooth, and in order do that we've introduced a few minor breaking changes to our experimental decorator behavior to make the behavior consistent with the standard decorator behavior.

Reviewer 2: Based on testing against Google's codebase, these changes impact about one in a thousand elements.

> The reason why we've included standard decorator support in a breaking change is that we've taken great care to make the upgrade path from experimental decorators smooth, and in order do that we've introduced a few minor breaking changes to our experimental decorator behavior to make the behavior consistent with the standard decorator behavior.

Reviewer 2: ```suggestion
The rewritten decorators behave slightly differently, just enough that the change is likely to break some users. Based on testing against Google's codebase, about one in a thousand elements will need changes. The primary change is that `@property` and `@state` will now call `requestUpdate` when applied to a setter function, where previously that was the setter's responsibility. If you have tests that didn't expect the element to update, they made need changes, or you may need to remove the decorator from the setter. This is rarely an issue in practice, as it should always be ok to call requestUpdate on an element.
```

> The big thing is that we've added support for decorating the new "auto-accessors" - fields prefixed with the `accessor` keyword. Auto-accessors create a getter/setter pair and private storage that they access. These are able to be overridden by standard decorators.

Reviewer 2: The big thing -> The most visible change is that

> Standard decorators _require_ that fields use the `accessor` keyword. To facilitate incremental upgrades to standard decorators, we now supportm them in experimental decorators, so that the same callsite works with both settings:

Reviewer 2: supportm -> support

> Standard decorators _require_ that fields use the `accessor` keyword. To facilitate incremental upgrades to standard decorators, we now supportm them in experimental decorators, so that the same callsite works with both settings:

Reviewer: ```suggestion
Standard decorators _require_ that fields use the `accessor` keyword. To facilitate incremental upgrades to standard decorators, we now support them in experimental decorators, so that the same call site works with both settings:
```

```ts
class MyElement extends LitElement {
  @property()
  accessor myProperty = 'hello';
}
```

> This means that you can upgrade your decorated fields to use `accessor`, and once they all are, you can turn off `experimentalDecorators`.

Reviewer: Can this line more explicitly state the enabling of incremental migration off experimental decorators to standard decorators, rather than the idea being implicit.

"This means that large projects using `experimentalDecorators` can incrementally migrate decorated fields to use the `accessor` keyword. Then when all usages use `accessor`, `experimentalDecorators` can be turned off without a change in behavior.

> The breaking changes were required because it's not possible to differentiate between auto-accessors and hand-written accessors. Previously we wouldn't automatically call `this.requestUpdate()` for accessors, instead requiring the hand-written setter to call it manually. Now we do call `requestUpdate()` automatically. We also will read the value of an accessor on first render and use that as the initial value for `changedProperties` and attribute reflection.

Reviewer: Since this is diving into some breaking changes, the change in `requestUpdate`, should other breaking changes be highlighted.

E.g. For custom property accessors (`@property`, or `@state`), decorators should now be set on the setter. This will be a breaking change for anyone who follows the recommendation on lit.dev.

import {compileLitTemplates} from '@lit-labs/compiler';

export default {
  // ...
  plugins: [
    typescript({
      transformers: {
        before: [compileLitTemplates()],
      },
    }),
    // other rollup plugins
  ],
};
```

## Preact Signals integration

Signals are all the rage right now. For Lit developers Signals promise to offer a convenient and relatively simple option for shared observable state.

One issue that signals present for web components is that signals lack interoperability between implementations. This goes against the interoperability goals of web components, so for now, rather than build our own signals package or pick just one that we endorse, we plan on offering integration packages to be able to use signal libraries with Lit.

These integrations will be relatively simple because we have two straight-forward ways of using signals:
1. Treat a component's update lifecycle as an effect, so that it's run when any signals it accesses (like in templates) changes. This integrates signals with Lit's batching async lifecycle for good performance when many signals change.
2. Use a directive to wire a signal directly to a location in DOM.

> The new `@lit-labs/preact-signals` package provides this integration for [Preact Signals](https://preactjs.com/guide/v10/signals/) which is published as a standalone library.

Reviewer: Same as compiler, can we add a prominent link somewhere to the npm package.

1. A `SignalWatcher` mixin, which makes a component automatically watch all signals used during updates.
2. A `watch()` directive, which watches one signal and updates a binding with it.
2. A special `html` template tag that auto-applies the `watch()` directive to all bindings that use signals.

Here's an example of using the `SignalWatcher` mixin:

```ts
import {LitElement, html} from 'lit';
import {customElement, property} from 'lit';
import {SignalWatcher, signal} from '@lit-labs/preact-signals';

const count = signal(0);

@customElement('signal-example')
export class SignalExample extends SignalWatcher(LitElement) {
  static styles = css`
    :host {
      display: block;
    }
  `;

  render() {
    return html`
      <p>The count is ${count.value}</p>
      <button @click=${this._onClick}>Increment<button></button></button>
    `;
  }

  private _onClick() {
    count.value = count.value + 1;
  }
}
```

## Lit Labs graduation

We are also graduating our first set of Lit Labs packages: Context, Task, and React.

These packages have a new home in the `@lit` npm scope, but are otherwise exactly the same as the current labs version.

> ##

Reviewer 2: Trailing header

### Breaking changes

- IE11 is no longer supported.
- Lit's npm modules are now published as ES2021.
- APIs deprecated with the Lit 2.0 release have been removed.
- SSR hydration support modules are moved to the `@lit-labs/ssr-client` package.
- Decorator behavior has been unified between TypeScript experimental decorators and standard decorators.
- Support was removed for Babel decorators version "2018-09"

These breaking changes should have minimal impact for most users – based on testing against Google's codebase, they impact about one in a thousand elements – and allow our project to move forward more efficiently.

### New: standard decorators support

The one new feature that we _did_ add to Lit 3.0 is support for the TC39 standard decorators specification to our _existing_ decorators.

The new decorator spec has reached Stage 3 in TC39, meaning that browsers and compilers are now implementing them. This is a huge step for Lit! The arrival of standard decorators allows us to begin the process of moving to a decorator implementation that won't require a compiler to use.

> It's very important to us to make the upgrade path from experimental decorators as smooth as possible. To accomplish this we made our existing decorators support the standard spec and made them work with the new `accessor` keyword in experimental decorator mode.

Reviewer: Nit: Maybe `It's very important __for__ us`? (instead of `to us to`)
